### DISCUSSION TITLE
Replace (instead of appending) HeadContent section from module

### DISCUSSION THREAD

**Discussion by mdmontesinos** [oqtane/oqtane.framework#4614] [Q&A] [answered]
I'm implementing a theme that has support to configure several SEO options with site and page theme settings. Therefore, the correspondent meta tags are added to the HeadContent with the ThemeBase method. However, there are some modules that need to override the default SEO tags added with their specific ones. For example, the Oqtane Blog module uses AddHeadContent for its meta tags.

In this case, you end up with duplicated seo meta tags, the ones from the theme and the ones from the module, as both of them are just appended to the site HeadContent. Is there any way to make the HeadContent behave as a dictionary (or similar) so that we can replace content by key instead of just appending?

**Comment by sbwalker** [comment, marked as answer]
@mdmontesinos this was explored previously but there are a number of significant challenges:

1. What is the dictionary key? Element type is not unique as there are multiple instances of most elements which are supported. Element "id" could be used however it would site administrators would need to specify the id property for each element - which I don't think is feasible (also html comments do not support id property). "Src" is supported for some elements but not others. 
2. Parsing the HtmlContent and BodyContent fields into individual elements is very challenging. And the more conditional logic required to try and distinguish the individual elements, the high probability it will impact performance as this operation much be executed on every page transition. 
3. Moving away from HtmlContent and BodyContent to use a different nack-end storage approach will still require a migration/upgrade effort to transform the old values into the new values - which requires complicated parsing (as identified in 2 above)

**Reply by mdmontesinos** [reply]
@sbwalker So what is the recommended approach for the issue I proposed? Or at least a "feasible" work-around, as being able to replace SEO meta tags depending on module data is quite critical for many applications. Should I just resort to js?

**Reply by sbwalker** [reply]
If it is the typical SEO metatags which you are trying to support then I would suggest that you do not specify them in the Site-level HeadContent, you should specify them in the Page-level HeadContent. And on the pages where you have a module which provides its own SEO metatags dynamically (ie. Blog) do not include them in the Page-level HeadContent for that specific page (as the module will inject them automatically).

**Reply by mdmontesinos** [reply]
That's actually not my case, as I'm not manually adding the tags in the HeadContent field. As I mentioned in my first message, I'm using theme settings to define them so the tags are added in the theme. I suppose I could add a new setting that instructs the theme to not render the site/page seo tags because a module will add them dynamically.

Anyway, thanks for the help!
